File: app/src/main/java/it/gruppoinfor/home2work/activities/SignInActivity.java
package it.gruppoinfor.home2work.activities;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.wang.avi.AVLoadingIndicatorView;

import java.net.UnknownHostException;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import es.dmoral.toasty.Toasty;
import it.gruppoinfor.home2work.App;
import it.gruppoinfor.home2work.R;
import it.gruppoinfor.home2work.utils.SessionManager;
import it.gruppoinfor.home2workapi.interfaces.LoginCallback;
import it.gruppoinfor.home2workapi.model.User;

public class SignInActivity extends AppCompatActivity implements LoginCallback {

    @BindView(R.id.sign_in_button)
    Button signInButton;
    @BindView(R.id.lost_password_button)
    TextView lostPasswordButton;
    @BindView(R.id.email_edit_text)
    EditText emailEditText;
    @BindView(R.id.password_edit_text)
    EditText passwordEditText;
    @BindView(R.id.sign_in_loading)
    AVLoadingIndicatorView signInLoading;

    private SessionManager mSessionManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sign_in);
        ButterKnife.bind(this);
        mSessionManager = new SessionManager(this);
    }

    @Override
    protected void onStart() {
        super.onStart();
        Bundle bundle = getIntent().getExtras();
        if (bundle != null) {

            switch (bundle.getInt(SessionManager.AUTH_CODE)) {
                case SessionManager.ERROR:
                    Toasty.error(this, "Errore durante il login", Toast.LENGTH_SHORT, true).show();
                    break;
                case SessionManager.EXPIRED_TOKEN:
                    Toasty.warning(this, getString(R.string.session_expired), Toast.LENGTH_SHORT, true).show();
                    break;
                case SessionManager.NO_INTERNET:
                    Toasty.error(this, "Nessuna connessione ad internet", Toast.LENGTH_SHORT, true).show();
                    break;
            }

        }

    }

    @Override
    public void onLoginSuccess() {
        mSessionManager.storeSession();

        User user = App.home2WorkClient.getUser();

        if (user.isConfigured()) {

            Intent i = new Intent(SignInActivity.this, MainActivity.class);
            startActivity(i);
            finish();

        } else {

            Intent i = new Intent(SignInActivity.this, ConfigurationActivity.class);
            startActivity(i);
            finish();

        }
    }

    @Override
    public void onInvalidCredential() {
        enableLogin();
        Toasty.error(SignInActivity.this, getString(R.string.login_failed), Toast.LENGTH_SHORT, true).show();
    }

    @Override
    public void onLoginError() {
        enableLogin();
        Toasty.error(SignInActivity.this, getString(R.string.server_error), Toast.LENGTH_SHORT, true).show();
    }

    @Override
    public void onError(Throwable throwable) {
        enableLogin();
        if (throwable instanceof UnknownHostException) {
            Toasty.error(SignInActivity.this, "Nessuna connessione ad internet", Toast.LENGTH_SHORT, true).show();
        } else {
            Toasty.error(SignInActivity.this, getString(R.string.server_error), Toast.LENGTH_SHORT, true).show();
        }
    }

    @OnClick(R.id.sign_in_button)
    void signInButtonClick() {

        if (!validate())
            return;

        emailEditText.setEnabled(false);
        passwordEditText.setEnabled(false);
        signInButton.setVisibility(View.GONE);
        signInLoading.setVisibility(View.VISIBLE);

        String email = emailEditText.getText().toString();
        String password = passwordEditText.getText().toString();

        App.home2WorkClient.login(email, password, false, this);

    }

    private void enableLogin() {
        signInButton.setVisibility(View.VISIBLE);
        emailEditText.setEnabled(true);
        passwordEditText.setEnabled(true);
    }

    private boolean validate() {

        boolean valid = true;

        String email = emailEditText.getText().toString();
        String password = passwordEditText.getText().toString();

        if (email.isEmpty() || !android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
            emailEditText.setError(getString(R.string.invalid_email_error));
            valid = false;
        } else {
            passwordEditText.setError(null);
        }

        if (password.isEmpty() || password.length() < 6) {
            passwordEditText.setError(getString(R.string.invalid_password_error));
            valid = false;
        } else {
            passwordEditText.setError(null);
        }

        return valid;
    }

}

File: App/src/main/java/it/gruppoinfor/home2work/activities/SignInActivity.kt
package it.gruppoinfor.home2work.activities

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Toast
import com.crashlytics.android.Crashlytics
import com.crashlytics.android.answers.Answers
import com.crashlytics.android.answers.LoginEvent
import com.pixplicity.easyprefs.library.Prefs
import it.gruppoinfor.home2work.R
import it.gruppoinfor.home2work.services.FirebaseTokenService
import it.gruppoinfor.home2work.user.SessionManager
import it.gruppoinfor.home2work.utils.Const
import it.gruppoinfor.home2workapi.HomeToWorkClient
import it.gruppoinfor.home2workapi.callback.LoginCallback
import kotlinx.android.synthetic.main.activity_sign_in.*
import java.net.UnknownHostException


class SignInActivity : AppCompatActivity(), LoginCallback {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_sign_in)

        sign_in_button.setOnClickListener {
            val view = this.currentFocus
            if (view != null) {
                val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                imm.hideSoftInputFromWindow(view.windowToken, 0)
            }

            if (validate()) {

                email_edit_text.isEnabled = false
                password_edit_text.isEnabled = false

                sign_in_button.visibility = View.INVISIBLE

                val email = email_edit_text.text.toString()
                val password = password_edit_text.text.toString()

                HomeToWorkClient.login(email, password, this)

            }
        }

        email_edit_text.setText(Prefs.getString(Const.PREFS_EMAIL, ""))
    }

    override fun onStart() {
        super.onStart()

        when (intent.extras?.getInt(Const.CODE_AUTH)) {
            Const.CODE_ERROR ->
                Toast.makeText(this, R.string.activity_signin_error, Toast.LENGTH_SHORT).show()
            Const.CODE_EXPIRED_TOKEN ->
                Toast.makeText(this, R.string.activity_signin_session_expired, Toast.LENGTH_SHORT).show()
            Const.CODE_NO_INTERNET ->
                Toast.makeText(this, R.string.activity_signin_no_internet, Toast.LENGTH_SHORT).show()
        }

    }

    override fun onLoginSuccess() {

        // Crashlytics log
        Answers.getInstance().logLogin(LoginEvent()
                .putMethod("Form")
                .putSuccess(true))

        // Salvo i dati dell'utente per crashalytics
        Crashlytics.setUserIdentifier(HomeToWorkClient.user?.id.toString())
        Crashlytics.setUserEmail(HomeToWorkClient.user?.email)
        Crashlytics.setUserName(HomeToWorkClient.user.toString())

        // Aggiorno il token Firebase Cloud Messaging sul server
        FirebaseTokenService().onTokenRefresh()

        Prefs.putString(Const.PREFS_EMAIL, HomeToWorkClient.user?.email)

        SessionManager.storeSession(this, HomeToWorkClient.user)

        val i = if (HomeToWorkClient.user!!.configured) {
            Intent(this@SignInActivity, MainActivity::class.java)
        } else {
            Intent(this@SignInActivity, ConfigurationActivity::class.java)
        }

        startActivity(i)
        finish()

    }

    override fun onInvalidCredential() {

        enableLogin()
        Toast.makeText(this@SignInActivity, R.string.activity_signin_login_failed, Toast.LENGTH_SHORT).show()

    }

    override fun onError(throwable: Throwable?) {

        // Crashlytics log
        Answers.getInstance().logLogin(LoginEvent()
                .putMethod("Form")
                .putSuccess(false))

        enableLogin()
        if (throwable is UnknownHostException) {
            Toast.makeText(this@SignInActivity, R.string.activity_signin_no_internet, Toast.LENGTH_SHORT).show()
        } else {
            Toast.makeText(this@SignInActivity, R.string.activity_signin_server_error, Toast.LENGTH_SHORT).show()
        }
    }

    private fun enableLogin() {

        sign_in_button.visibility = View.VISIBLE
        email_edit_text.isEnabled = true
        password_edit_text.isEnabled = true

    }

    private fun validate(): Boolean {

        var valid = true

        val email = email_edit_text.text.toString()
        val password = password_edit_text.text.toString()

        if (email.isEmpty() || !android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
            email_edit_text.error = getString(R.string.activity_signin_email_error)
            valid = false
        } else {
            password_edit_text.error = null
        }

        if (password.isEmpty() || password.length < 6) {
            password_edit_text.error = getString(R.string.activity_signin_password_error)
            valid = false
        } else {
            password_edit_text.error = null
        }

        return valid
    }



}
